[Feat] 밴드 정보 입력 VC에서 합주곡을 추가하는 기능을 구현했습니다.

Rookie0031 opened the pull request:
## 관련 이슈
- closed #157 

## 배경
 밴드 생성 시 합주곡을 추가하기 위해서는 아래와 같은 카드 형태로 정보를 입력합니다. 이 UI를 구현했습니다.
<img width="400" alt="image" src="https://user-images.githubusercontent.com/103009135/224059593-a8f364b7-7037-4b7a-94f1-424f0f852154.png">
</br>

## 테스트 방법
시작점을 아래와 같이 바꿔주시고 합주곡이 잘 추가되나 확인해주세요!

        window.rootViewController = UINavigationController(rootViewController: BandInformationSetViewController())


## 작업 내용

### 1️⃣ 합주곡 추가 레이아웃 구조 (AddPracticeSongViewController)

#### 이 VC 레이아웃은 MainScrollView아래 UIStackView로 ContentView를 잡은 구조입니다.

ContentView의 arrangedSubview는 PracticeSongCardView라는 UIStackView로, 유저가 합주곡 정보를 입력할 수 있는 3개의 텍스트필드가 들어있습니다.

테이블뷰나 콜렉션뷰를 사용하지 않고 스크롤뷰를 채택한 이유는 좀 더 빠른 개발을 위해서입니다.
이 VC의 UI에는 합주곡 추가 버튼이 항상 카드뷰의 하단에 배치되게 됩니다. 그럼 셀의 추가에 따라 이 버튼을 고정시키는 레이아웃을 구현하기가 어렵다고 판단했습니다. 또한 이 기능을 개발하는 당시 DiffableDataSource에 익숙치않은 상황이었습니다.

<img width="666" alt="image" src="https://user-images.githubusercontent.com/103009135/223443412-183862e2-e618-4125-8c03-576f0645da89.png">
위 그림에서 합주곡 추가 버튼이 보라색인데, 실제는 회색입니다. 예전에 미리 작성된 PR 내용을 가져와서 오류가 있네요 ㅎㅎ..
</br>

### 2️⃣ 추가 합주곡 전달 구조

이렇게 생성된 정보는 밴드정보 입력 VC로 클로저를 통해 전달됩니다.
이전의 노엘이 구현해준 CollectionViewCell의 UI 정보로 PracticeSongBoxView 를 구현했고 전달된 정보를 바탕으로 stackView를 configure하고, configured된 stackView를 삽입하는 구조입니다.

<img width="660" alt="image" src="https://user-images.githubusercontent.com/103009135/223445078-0337e522-ffa0-48dc-a02a-592ba57ed622.png">

https://github.com/extreme-rock/GetARock-iOS/blob/e210ea71582b6e6c2a8a5c12cc44a892c65e8f85/GetARock/GetARock/Screens/BandCreation/BandInformationSetFlow/BandInformationSetViewController.swift#L307-L318
</br>

### 3️⃣ 추가된 합주곡 및 합주곡 입력 카드 삭제 시 애니메이션 구현

https://user-images.githubusercontent.com/103009135/224062701-d70d426a-a98f-4eb5-94ca-04b67614f7b2.MP4

 UIStackView를 사용하면 View를 삭제할 때 removeFromSuperView를 사용하면 되기 때문에 삭제 코드가 간단하고 삭제에 따른 레이아웃 업데이트가 쉽습니다. 하지만 애니메이션이 없어 UX가 저해되었습니다. 그래서 최대한 이를 메꾸고자 UIView animation을 넣어보았습니다.
</br>

https://github.com/extreme-rock/GetARock-iOS/blob/e210ea71582b6e6c2a8a5c12cc44a892c65e8f85/GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift#L186-L196
</br>


### 4️⃣ 필수 정보 누락에 따른 최종 추가 버튼 비활성화 기능
#### 1. 각 카드뷰의 텍스트필드는 텍스트가 변할 때마다 Notificaiton을 날립니다. 
https://github.com/extreme-rock/GetARock-iOS/blob/e210ea71582b6e6c2a8a5c12cc44a892c65e8f85/GetARock/GetARock/Global/UIComponent/BasicTextField.swift#L9-L11

https://github.com/extreme-rock/GetARock-iOS/blob/e210ea71582b6e6c2a8a5c12cc44a892c65e8f85/GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/PracticeSongCardView.swift#L121-L125


2. AddPracticeSongViewController 에서 Notification을 수신합니다.
https://github.com/extreme-rock/GetARock-iOS/blob/e210ea71582b6e6c2a8a5c12cc44a892c65e8f85/GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift#L156-L161


3. contentStackView의 하위뷰들 (PracticeCardView)에 접근해서 모든 필수정보가 입력되었는지 체크합니다. 그 체크 여부에 따라 버튼이 활성, 비활성화 됩니다.

https://github.com/extreme-rock/GetARock-iOS/blob/e210ea71582b6e6c2a8a5c12cc44a892c65e8f85/GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift#L217-L229
</br>

### 🏃다음 PR 
####  이 PR이 승인되면 전달받은 정보들을 모두 수합해서 POST를 날리는 함수가 구현될 예정입니다. 
</br>

## 리뷰 노트
- UI는 작아보이지만 세부적인 내용이 많아 UI Testing에서 버그가 있을지 궁금합니다
- 불필요한 코드 및 부적절한 변수명
- 컨벤션 미준수 사항
- 로직 개선 가능 사항


<!-- 해당 PR 리뷰 과정에서 논의된 확정 사항을 develop 브랜치에 머지하기 전 정리합니다. -->

Rookie0031 committed:
[Design] 합주곡 추가 Card, Box UI 추가 (#157)

1. PracticeSongCardView : 합주곡에 관한 정보를 입력할 때, 카드 형태의 정보 입력 UI
2. PracticeSongBoxView: 합주곡을 정보 입력 뷰에서 Box 형태로 추가되는 UI

Rookie0031 committed:
[Feat] 합주곡 입력 정보로 전달 구조 추가 (#157)

Rookie0031 committed:
[Design] 합주곡 추가 VC에서 합주곡 카드 추가 버튼 비활성화 구조 구현 (#157)

Rookie0031 committed:
[Feat] 필수 정보 미입력에 따른 추가 버튼 활성/비활성 여부 체크 기능 구현 (#157)

Rookie0031 committed:
[Del] 불필요한 코드 삭제 (#157)

Rookie0031 committed:
[Design] 네비게이션 타이틀 추가 (#157)

Rookie0031 commented on GetARock/GetARock/Screens/BandCreation/BandInformationSetFlow/BandInformationSetViewController.swift:
```diff
@@ -279,6 +279,15 @@ extension BandInformationSetViewController {
     //MARK: 합주곡 추가 기능 관련 로직
     @objc func didTapAddPracticeSong() {
+        let nextViewController = AddPracticeSongViewController()
+        nextViewController.completion = { [weak self] songs in
```
유저가 입력한 합주곡 정보를 받아서 StackView에 주입하는 과정입니다

Rookie0031 commented on GetARock/GetARock/Screens/BandCreation/BandInformationSetFlow/BandInformationSetViewController.swift:
```diff
@@ -279,6 +279,15 @@ extension BandInformationSetViewController {
+        nextViewController.completion = { [weak self] songs in
+            let addedSongs: [PracticeSongBoxView] = self?.makePracticeSongBoxes(with: songs) ?? []
+            for song in addedSongs {
+                if self?.practiceSongList.arrangedSubviews.count ?? 0 > 3 { break }
```
합주곡이 3개 이상 추가되는것을 막기 위해 break로 박스 형태의 합주곡 UI를 추가하는 for문을 탈출합니다

Rookie0031 commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift:
```diff
@@ -0,0 +1,248 @@
+                                     padding: UIEdgeInsets(top: 40, left: 20, bottom: 10, right: 20))
+    }
+    
+    private func updateDeleteButtonState() {
```
만약 합주곡 정보 입력 카드가 1개가 될 경우 삭제 버튼이 없어지는 로직입니다

Rookie0031 commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift:
```diff
@@ -0,0 +1,248 @@
+        )
+    }
+    
+    private func deleteFirstPracticeSongCard() {
```
첫번째 카드는 VC에 들어오는 순간 init 되서 존재하기 때문에 따로 삭제 action을 넣어줘야했습니다

Rookie0031 commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift:
```diff
@@ -0,0 +1,248 @@
+    func didTapAddPracticeSong() {
+        guard contentView.arrangedSubviews.count < 3 else { return }
+        let newCard = PracticeSongCardView()
+        newCard.setTextFieldDelegate(controller: self)
```
합주곡 정보 입력에는 곡이름, 가수, 링크 정보로 3개의 텍스트필드가 들어갑니다. 텍스트 필드에 텍스트가 입력되면 스크롤 상태에서는 키보드를 올려야하기 때문에 델리게이트 설정이 필요했습니다

ccdkk commented on GetARock/GetARock/Global/UIComponent/BasicTextField.swift:
```diff
@@ -6,27 +6,42 @@
 import UIKit
 
+protocol BasicTextFieldDelegate: AnyObject {
+    func textFieldTextDidChange()
```
p3;
changed는 어떤가요?

Rookie0031 replied:
뭔가 이게 더 애플 API 이름이랑 비슷해서 썻는데 ex. textViewDidChange(_:)
 changed가 더 직관적이신가요??

ccdkk replied:
오 그러네요 change 굳굳!

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift:
```diff
@@ -0,0 +1,248 @@
+                addPracticeSongButton.isHidden = true
+            } else {
+                addPracticeSongButton.isHidden = false
+            }
```
p4;
삼항연산자로 나타낼 수 있을 것 같아요!
```suggestion
                addPracticeSongButton.isHidden = numberOfSong == 3 ? true : false
```

Rookie0031 replied:
감사합니다 반영완료!! https://github.com/extreme-rock/GetARock-iOS/pull/158/commits/f8c0504f5e97380a7901985981280938a6d4ace5

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift:
```diff
@@ -0,0 +1,248 @@
+            self?.deleteFirstPracticeSongCard()
+        }
+        firstPracticeSongCard.deleteButton.addAction(deleteAction, for: .touchUpInside)
+        return $0
```
q;
firstPracticeSongCard의 deleteButton addAction을 여기서 하시는 이유가 있을까요?

Rookie0031 replied:
헉 왜 여기 들어가있을까요 firstPracticeSongCard로 옮겼습니다 감사합니다! https://github.com/extreme-rock/GetARock-iOS/pull/158/commits/cddd0b297cbbc519c8dd2e9bc092732d54b1fcb0

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift:
```diff
@@ -0,0 +1,248 @@
+    }(UIStackView(arrangedSubviews: [firstPracticeSongCard]))
+    
+    private lazy var mainScrollView: UIScrollView = {
+        $0.showsVerticalScrollIndicator = true
```
p3;
default값이 true라 굳이 써줄 필요 없을 것 같아요!

Rookie0031 replied:
오옷 넵 수정완료!

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift:
```diff
@@ -0,0 +1,248 @@
+        // contentView는 stackView라서 높이정보가 자동으로 세팅됨
+        // 따라서 스크롤범위를 맞추기위해 contentView의 하단 constraint가 필요함
+        contentView.constraint(top: mainScrollView.contentLayoutGuide.topAnchor,
+                               bottom: mainScrollView.contentLayoutGuide.bottomAnchor, centerX: mainScrollView.centerXAnchor,
```
p3;
개행 부탁드려요~

요 부분에서 mainScrollView contentLayoutGuide의 leading과 trailing이 모호해서 hierarchy를 확인해보면 보라색 경고창이 뜨네요. 그래서 centerX를 없애고 contentView의 leading과 trailing을 각각 contentLayoutGuide의 leading, trailing에 맞췄는데
고렇게 하니까 BasicTextField에 있는 leftPaddingView(?) 에서 보라색 에러가 뜨네요 ㅜㅜ 그래서 뭔가 수정버전까지 코멘트로 올리려고 했는데 일단 요 상황까지 올리겠읍니다..
<img width="628" alt="스크린샷 2023-03-13 오후 10 46 09" src="https://user-images.githubusercontent.com/91456952/224720342-d38c0ad7-b65a-4268-8a07-8adc687bd078.png">

Rookie0031 replied:
오 저런 문제가 있었군요 감사합니다!! 언급해주신 부분은 고쳤습니다! https://github.com/extreme-rock/GetARock-iOS/pull/158/commits/911a61cc1718b3d2c2186b3f45233cf365680cc5

### TextField PaddingView 삽입시 Position ambiguous 문제

 처음에 TextField에 leftview 프로퍼티는 위치가 고정된 view라서 UIView의 크기 높이만 지정해주면 되는 줄 알았는데요, 이 친구도 마찬가지로 위치에 대한 constraint이 필요해서 발생된 문제였습니다.
그래서 leftView와 rightView에 패딩뷰 삽입시 Frame을 사용해서 위치와 크기 정보를 같이 주니까 해결되었습니다.

            // 변경 전
            $0.leftView = TextFieldLeftPaddingView( )
                       
            // 변경 사항 후
            $0.leftView = UIView(frame: CGRect(x: 0, y: 0, width: 20, height: 20))


텍스트필드는 왼쪽에 기본적인 패딩값이 세팅되어야하기 때문에 위와 같이 설정부에서 바로 UIView를 선언하고 삽입해주면 되서 TextFieldLeftPaddingView는 삭제되었습니다!

우측패딩값을 주는 TextFieldRightPaddingView는 상황에 따라 크기 변화가 필요하기 때문에 클래스를 남겨 두었고 init 과정에서 Frame을 기본 설정해놓았습니다

              final class TextFieldRightPaddingView: UIView {
                  override init(frame: CGRect) {
                      super.init(frame: CGRect(x: 0, y: 0, width: 100, height: 20))
                  }


### AddPracticeScrollView에서 contentView Layout ambiguous 문제
 말씀하신 대로 contentLayoutGuide에 leading trailing을 맞춰보니 이상하게 contentView가 중앙정렬 되지않는 문제가 있었습니다. (contentView의 alignment를 center로 지정해도 같은 문제 발생)

   <img width="200" alt="스크린샷1" src="https://user-images.githubusercontent.com/103009135/224951568-2e0d94d0-932e-4991-b1e0-6f2416b1958e.png">


그래서 그냥 contentLayoutGuide를 이용하지 않으니 문제가 해결되었습니다. (왜인지는 아직 모르겟네요..)

        contentView.constraint(
            top: mainScrollView.topAnchor,
            leading: mainScrollView.leadingAnchor,
            bottom: mainScrollView.bottomAnchor,
            trailing: mainScrollView.trailingAnchor,
            padding: UIEdgeInsets(top: 20, left: 16, bottom: 160, right: ### 16))
 

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/103009135/224952950-e7714185-08e4-47be-b2b4-d87ac148e340.png">

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift:
```diff
@@ -0,0 +1,248 @@
+            contentView.arrangedSubviews.compactMap { $0 as? PracticeSongCardView }.forEach { $0.deleteButton.isHidden = true }
+        } else {
+            contentView.arrangedSubviews.compactMap { $0 as? PracticeSongCardView }.forEach { $0.deleteButton.isHidden = false }
+        }
```
p3;
반복되는 부분을 앞으로 빼면 좀더 간결하게 만들 수 있을 것 같습니다!
```Swift
let practiceSongs = contentView.arrangedSubviews.compactMap { $0 as? PracticeSongCardView }
        if contentView.arrangedSubviews.count == 1 {
            practiceSongs.forEach { $0.deleteButton.isHidden = true }
        } else {
            practiceSongs.forEach { $0.deleteButton.isHidden = false }
        }
```

그리고 deleteButton에 직접 접근하는 것보다 isHidden을 설정해주는 함수를 만들어서 적용해보는 것은 어떨까용?

Rookie0031 replied:
오옷 감사합니다!! 반영완료! https://github.com/extreme-rock/GetARock-iOS/pull/158/commits/cddd0b297cbbc519c8dd2e9bc092732d54b1fcb0

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift:
```diff
@@ -0,0 +1,248 @@
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(getKeyboardHeight(notification: )),
+            name: UIResponder.keyboardWillShowNotification, object: nil
```
p3;
object에 개행 부탁드려요!

Rookie0031 replied:
넵 반영완료!

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift:
```diff
@@ -0,0 +1,248 @@
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(updateCompleteButtonState),
+            name: Notification.Name.didPracticeCardViewTextFieldChange, object: nil
```
p3;
object 개행 부탁드려요!

Rookie0031 replied:
넵 반영완료!

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift:
```diff
@@ -0,0 +1,248 @@
+            })
+        }
+        newCard.deleteButton.addAction(deleteAction, for: .touchUpInside)
+        
```
p3;
deleteButton을 action과 함께 만들어두고 isHidden으로 관리하신게 아니라 뉴카드 생길 때마다 따로 만들어주시는 이유가 있을까요?

Rookie0031 replied:
급하게 작업하다보니 가독성이 좀 안좋은 부분이있었습니다 하하.. 
그런데 "deleteButton을 action과 함께 만들어두고 isHidden으로 관리한다" 라는게 어떤 의미인가요??

        let deleteAction: UIAction = UIAction { [weak self] _ in
            UIView.animate(withDuration: 0.4, animations: {
                newCard.alpha = 0 // fade out 애니메이션
            }, completion: { [weak self] _ in
                newCard.removeFromSuperview()
                self?.numberOfSong = self?.contentView.arrangedSubviews.count ?? 0
                UIView.animate(withDuration: 0.3) { // StackView 레이아웃 재조정 애니메이션
                    self?.contentView.layoutIfNeeded()
                }
            })
        }

요 부분이 어짜피 중복되니까 따로 프로퍼티로 빼서 관리하면 좋다는 말인가요???
삭제 로직에서 AddPracticeSongVC의 contentView에 접근하기 때문에 PracticeCardView에서 삭제로직을 넣지 못햇는데 escaping closure 로 넣으면 될 거 같기도 하네요,,

ccdkk replied:
아하 맞아요. 정확하십니다. 그런 이유에서라면 그래도 괜찮을 것 같네요.

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/AddPracticeSongViewController.swift:
```diff
@@ -0,0 +1,248 @@
+        } else {
+            addCompleteButton.isEnabled = false
+        }
+    }
```
p2;
이 친구를 cell을 delete할 때도 체크해줘야할 것 같아요. 조건이 만족된 다음, 새로운 cell을 추가하고 해당 cell을 삭제하면, button이 비활성화되어 있더라구요.
![Simulator Screen Recording - iPhone 13 Pro - 2023-03-13 at 22 22 03](https://user-images.githubusercontent.com/91456952/224725824-954b827d-a652-4b6b-80b7-82d4d89aed03.gif)

Rookie0031 replied:
큰일날뻔 햇네요 쏘 굿... 감사합니다 반영햇어요!!  https://github.com/extreme-rock/GetARock-iOS/pull/158/commits/cddd0b297cbbc519c8dd2e9bc092732d54b1fcb0

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/PracticeSongBoxView.swift:
```diff
@@ -0,0 +1,104 @@
+        let imageConfiguation = UIImage.SymbolConfiguration(pointSize: 20)
+        $0.setImage(UIImage(
+            systemName: "xmark.circle.fill",
+            withConfiguration: imageConfiguation),for: .normal
```
p3;
개행 부탁드려요!

Rookie0031 replied:
넵 반영완료! https://github.com/extreme-rock/GetARock-iOS/pull/158/commits/cddd0b297cbbc519c8dd2e9bc092732d54b1fcb0

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/PracticeSongBoxView.swift:
```diff
@@ -0,0 +1,104 @@
+        )
+        $0.contentMode = .scaleAspectFit
+        $0.tintColor = .gray02
+        let action = UIAction { [weak self] _ in
+            UIView.animate(withDuration: 0.3, animations: {
+                self?.alpha = 0
+            }, completion: { [weak self] _ in
+                self?.removeFromSuperview()
+            })
+        }
```
요렇게 fadeout 구현하는거 신기하네요!!!

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/PracticeSongCardView.swift:
```diff
@@ -0,0 +1,125 @@
+    func getSongName() -> String {
+        return practiceSongTextField.textField.text ?? ""
+    }
+    
```
p3;
swift에서는 get을 사용한 네이밍을 지양하더라구요. 그냥 artistName()과 songName()은 어떤가요?

Rookie0031 replied:
쏘 굿입니다.. 반영 완료!

ccdkk commented on GetARock/GetARock/Screens/BandCreation/PracticeSongAddFlow/PracticeSongCardView.swift:
```diff
@@ -0,0 +1,125 @@
+        return practiceSongTextField.textField.text ?? ""
+    }
+    
+    func setTextFieldDelegate(controller: UITextFieldDelegate) {
+        artistNameTextField.textField.delegate = controller
+        practiceSongTextField.textField.delegate = controller
+        linkTextField.textField.delegate = controller
+    }
```
q;
오옷 요거는 어떤 구조 인가요?

Rookie0031 replied:
AddPracticeSongVC 에서 텍스트 필드 입력에 따라 키보드가 올라가야하는 것을 구현하기 위한 것입니다.
VC의 프레임을 위로 올려서 키보드가 올라옴에 따른 정보 입력뷰가 가려지는 것을 방지하는 구조인데 이 VC를 여러개의 TextField 의 델리게이트로 설정하기 위해 만든 함수입니다!

      extension AddPracticeSongViewController: UITextFieldDelegate {
          func textFieldDidBeginEditing(_ textField: UITextField) {
              self.view.frame.origin.y -= self.keyBoardHeight
          }
          
          func textFieldDidEndEditing(_ textField: UITextField) {
              self.view.frame.origin.y += self.keyBoardHeight
          }
      }


    func didTapAddPracticeSong() {
        guard contentView.arrangedSubviews.count < 3 else { return }
        let newCard = PracticeSongCardView()
        newCard.setTextFieldDelegate(controller: self) // 이 부분에서 사용됨

ccdkk reviewed (commented):
고생하셨습니다!

Rookie0031 committed:
[Style] 합주곡 추가 버튼 숨김처리에서 삼항연산자 적용 (#157)

데이크의 PR 리뷰 사항 반영

Co-authored-by: dake <cdk243@gmail.com>

Rookie0031 committed:
[Fix] 세부 UI 레이아웃 관련 warning 사항 수정 (#157)

Rookie0031 committed:
[Fix] 합주곡 추가 완료 버튼 상태 변경 관련 버그 수정 및 변수명 수정 (#157)

https://github.com/extreme-rock/GetARock-iOS/pull/158/files#r1134008083

PR 리뷰 참고

Rookie0031 committed:
Merge branch 'develop' into feature/157-add-practice-songs
